addons/mail/sfu_server/src/shared/bus.js

```diff
@@ -0,0 +1,159 @@
+/**
+ * Bus class that implements a request/response pattern on top of a websocket.
+ *
+ * Compatible in both Node and Browser environment.
+ */
+export class Bus {
+    /** @type {Map<number, Bus>} */
+    static records = new Map();
+    /**
+     * Used to know if the code runs on the client to avoid ID collisions as this class can is defined in both
+     * the server(/node) and the client(/browser) environments (and therefore not share static properties).
+     *
+     * @type {"c"|"s"}
+     */
+    static _type =
+        typeof window !== "undefined" && typeof window.document !== "undefined" ? "c" : "s";
+    /** @type {number} */
+    static _idCount = 0;
+
+    /** @type {Function} */
+    onClose;
+    /** @type {Function} */
+    onMessage;
+    /** @type {Function} */
+    onRequest;
```

Comment:
related to https://github.com/odoo/odoo/pull/132153#discussion_r1358168366 , it avoided mixing `EventTarget` and `EventEmitter` server-side and I could use the callback return value instead of passing a callback to respond. I will keep experimenting with alternative Bus API when I have some time

Reply:
Node actually implements `EventTarget`, so while you can't really get around a compat layer for things like `WebSocket` which are one or the other depending on the implementation, for custom code you can just choose to use `EventTarget`. You can also pretty easily override addEventListener to call a `detail.setResult` callback automatically, and create a new method like `trigger` or `emit` that just takes an event type and detail, and handles constructing the event, giving it the `setResult` callback and returning the result that was set, this really simplifies the API on the caller side, something like
```js
class EventBus extends EventTarget {
  addEventListener(eventType, cb, options) {
    const wrappedCb = (ev) => {
      ev.detail.setResult(cb(ev));
    };
    return super.addEventListener(eventType, wrappedCb, options);
  }
  trigger(eventType, detail) {
    let result;
    this.dispatchEvent(
      new CustomEvent(eventType, {
        detail: {
          ...detail,
          setResult: (res) => {
            result = res;
          },
        },
      })
    );
    return result;
  }
}
```
If you don't want to have to destructure the details out of the event in every callback you could also do that in the wrapped callback and call the callback directly with details, though typescript might get mad that the API is not compatible.